modules/xfem/src/userobjects/InterfaceMeshCutUserObjectBase.C

```diff
@@ -0,0 +1,199 @@
+      mooseError("UserObject casting to XFEMMovingInterfaceVelocityBase in "
+                 "MovingLineSegmentCutSetUserObject");
+
+    _interface_velocity = dynamic_cast<const XFEMMovingInterfaceVelocityBase *>(uo);
+    const_cast<XFEMMovingInterfaceVelocityBase *>(_interface_velocity)->initialize();
+  }
+
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+    _initial_nodes_location[node->id()] = *node;
+
+  for (const auto & elem : _cutter_mesh->element_ptr_range())
+    for (unsigned int n = 0; n < elem->n_nodes(); n++)
+      _node_to_elem_map[elem->node_id(n)].push_back(elem->id());
+
+  _cutter_mesh->prepare_for_use();
+  _cutter_mesh->set_mesh_dimension(_mesh.dimension() - 1);
+
+  if (_output_exodus)
+  {
+    _equation_systems = libmesh_make_unique<EquationSystems>(*_cutter_mesh);
+    _explicit_system = &(_equation_systems->add_system<ExplicitSystem>("InterfaceMeshSystem"));
+
+    _explicit_system->add_variable("disp_x");
+    _explicit_system->add_variable("disp_y");
+
+    if (_mesh.dimension() == 3)
+      _explicit_system->add_variable("disp_z");
+
+    _equation_systems->init();
+    _exodus_io->write_equation_systems(_app.getOutputFileBase() + "_" + name() + ".e",
+                                       *_equation_systems);
+  }
+}
+
+void
+InterfaceMeshCutUserObjectBase::initialize()
+{
+  std::vector<Point> new_position(_cutter_mesh->n_nodes());
+
+  _pl = _mesh.getPointLocator();
+  _pl->enable_out_of_mesh_mode();
+
+  std::map<unsigned int, Real> node_velocity;
+  Real sum = 0.0;
+  unsigned count = 0;
+  // Loop all the nodes to calculate the velocity
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+  {
+    if ((*_pl)(*node) != nullptr)
+    {
+      Real velocity;
+      if (_func == nullptr)
+        velocity =
+            _interface_velocity->computeMovingInterfaceVelocity(node->id(), nodeNormal(node->id()));
+      else
+        velocity = _func->value(_t, *node);
+
+      // only updates when t_step >0
+      if (_t_step <= 0)
+        velocity = 0.0;
+
+      node_velocity[node->id()] = velocity;
+      sum += velocity;
+      count++;
+    }
+  }
+
+  if (count == 0)
+    mooseError("No node of the cutter mesh is found inside the computational domain.");
+
+  Real average_velocity = sum / count;
+
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+  {
+    if ((*_pl)(*node) == nullptr)
+      node_velocity[node->id()] = average_velocity;
+  }
+
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+  {
+    Point p = *node;
+    p += _dt * nodeNormal(node->id()) * node_velocity[node->id()];
+    new_position[node->id()] = p;
+  }
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+    _cutter_mesh->node_ref(node->id()) = new_position[node->id()];
+
+  if (_output_exodus)
+  {
+    std::vector<dof_id_type> di;
+    for (const auto & node : _cutter_mesh->node_ptr_range())
+    {
+      _explicit_system->get_dof_map().dof_indices(
```

Comment:
If you had a member variable that stores the variable numbers for the displacements, you could make this code more compact by looping through that. I suspect that doing these calls to `variable_number()` with strings for every node is expensive, so it would make the code faster too.